### PR TITLE
enhancement: adding filtered applications count 

### PR DESCRIPTION
**Pull request by rishirishhh**
Date: 23/09/2024

Developer Name: Kavya Rishi 

---

## Issue Ticket Number

Closes: https://github.com/Real-Dev-Squad/website-dashboard/issues/825

## Description

A new functionality was added to count the total number of applications that match the given status.
This was achieved by introducing a countQuery, which fetches all documents that match the status

### Documentation Updated?

- [ ] Yes
- [X] No

### Under Feature Flag

- [X] Yes
- [ ] No

### Database Changes

- [ ] Yes
- [X] No

### Breaking Changes

- [ ] Yes
- [X] No

### Development Tested?

- [x] Yes
- [ ] No

## Screenshots
<details>
<summary>Screenshot 1</summary>

Screenshot from 2024-09-23 01-56-39
Screenshot from 2024-09-23 01-56-51

</details>

## Test Coverage
<details>
<summary>Screenshot 1</summary>
Added them below

</details>

### CODE REVIEW

**Comment by rishirishhh** [comment]
Screenshot from 2024-09-23 01-43-34

**Comment by rishirishhh** [comment]
Screenshot from 2024-09-23 01-43-57
Screenshot from 2024-09-23 01-44-14

File: controllers/applications.ts [hunk 1]
```diff
@@ -9,9 +9,10 @@ const admin = require("firebase-admin");
 
 const getAllOrUserApplication = async (req: CustomRequest, res: CustomResponse): Promise<any> => {
   try {
-    const { userId, status, next, size } = req.query;
+    const { userId, status, next, size, dev } = req.query;
     const limit = Number(size) || 25;
     let nextPageUrl = null;
+    const isDevMode = dev === "true"; // Check if the 'dev' query param is true
```
Comment (vinit717) [inline controllers/applications.ts:15]: Remove the comment

File: controllers/applications.ts [hunk 2]
```diff
@@ -23,17 +24,27 @@ const getAllOrUserApplication = async (req: CustomRequest, res: CustomResponse):
...
+        limit,
+        next
+      );
 
       if (applications.length === limit) {
         nextPageUrl = `/applications?next=${lastDocId}&size=${limit}&status=${status}`;
       }
 
-      return res.json({
+      // Include totalCount only if 'dev' is true
```
Comment (vinit717) [inline controllers/applications.ts:37]: Remove comment

File: test/unit/middlewares/application-validator.test.ts [hunk 3]
```diff
@@ -40,10 +40,10 @@ describe("application validator test", function () {
...
 
     it("should not call the next function if any of the values which have a wordCount restriction doesn't contain the expected number of words", async function () {
-       const rawData = {
-         ...applicationsData[6],
-         whyRds: 'jfaskdfjsd'
-       };
+      const rawData = {
+        ...applicationsData[6],
+        whyRds: "jfaskdfjsd",
+      };
```
Comment (vinit717) [inline test/unit/middlewares/application-validator.test.ts:46]: Why this changed?

File: test/unit/middlewares/application-validator.test.ts [hunk 4]
```diff
@@ -61,10 +61,10 @@ describe("application validator test", function () {
...
 
     it("should not call the next function if number of hours is not a number", async function () {
-       const rawData = {
-         ...applicationsData[6],
-         numberOfHours: '10'
-       };
+      const rawData = {
+        ...applicationsData[6],
+        numberOfHours: "10",
+      };
```
Comment (vinit717) [inline test/unit/middlewares/application-validator.test.ts:67]: Here also

File: controllers/applications.ts
Comment (vinit717) [inline controllers/applications.ts:1]: Write integration test

File: models/applications.ts
Comment (vinit717) [inline models/applications.ts:1]: Remove unnecessary code changes and comments

**Review by vinit717 — CHANGES_REQUESTED** [review]

File: test/unit/middlewares/application-validator.test.ts (hunk 3 above)
Comment (rishirishhh) [reply test/unit/middlewares/application-validator.test.ts:46]: They were done by lint in my vscode system

File: test/unit/middlewares/application-validator.test.ts (hunk 4 above)
Comment (rishirishhh) [reply test/unit/middlewares/application-validator.test.ts:67]: They were done by lint in my vscode system

File: controllers/applications.ts
Comment (rishirishhh) [reply controllers/applications.ts:1]: I have written them

File: test/integration/application.test.ts [hunk 5]
```diff
@@ -206,6 +206,29 @@ describe("Application", function () {
         });
     });
 
+    it("should return application with status rejected and the total count of the rejected applications if status rejected is passed in query params with dev = true and next url if the size provided is equal to the applications returned in query ", function (done) {
```
Comment (vinit717) [inline test/integration/application.test.ts:209]: Can you make this title small

File: test/integration/application.test.ts
Comment (vinit717) [inline test/integration/application.test.ts:1]: What about negative test case

File: test/unit/middlewares/application-validator.test.ts (hunk 3 above)
Comment (vinit717) [reply test/unit/middlewares/application-validator.test.ts:46]: so remove this changes

File: models/applications.ts [hunk 6]
```diff
@@ -81,7 +81,14 @@ const getApplicationsBasedOnStatus = async (status: string, limit: number, lastD
       });
     });
 
-    return { applications, lastDocId: lastApplicationDoc?.id };
+    // Count total number of applications matching the status and optional userId
+    let countQuery = ApplicationsModel.where("status", "==", status);
+
+    // Get the total count of applications matching the criteria
+    const totalApplications = await countQuery.get();
+    const totalCount = totalApplications.size; // Get the number of documents in the snapshot
```
Comment (vinit717) [inline models/applications.ts:89]: remove the comments

File: test/integration/application.test.ts
Comment (rishirishhh) [reply test/integration/application.test.ts:1]: Yes I have added a negative case where lets say if the api does not have the dev flag present then the response given by the API should not contain totalCount property, hence you will not see the count of the applications of selected status.
I have added that in a already existing test.

File: test/integration/application.test.ts (hunk 5 above)
Comment (rishirishhh) [reply test/integration/application.test.ts:209]: Done

**Review by Ayushsanjdev — APPROVED** [review]
LGTM!

**Review by Achintya-Chatterjee — DISMISSED** [review]

File: test/integration/application.test.ts [hunk 7]
```diff
@@ -202,6 +202,30 @@ describe("Application", function () {
           expect(res.body.next).to.be.equal(
             `/applications?next=${res.body.applications[res.body.applications.length - 1].id}&size=2&status=rejected`
           );
+          expect(res.body).to.not.have.property("totalCount");
+          return done();
+        });
+    });
+
+    it("should return application with status rejected and the total count of the rejected applications if  dev = true ", function (done) {
```
Comment (vinit717) [inline test/integration/application.test.ts:210]: What about all and pending status?

**Review by shubhdevelop — DISMISSED** [review]
Looks Good to me

File: test/integration/application.test.ts (hunk 7 above)
Comment (rishirishhh) [reply test/integration/application.test.ts:251]: I have added them in the new commit

**Review by Achintya-Chatterjee — APPROVED** [review]

**Review by Achintya-Chatterjee — APPROVED** [review]

**Review by shubhdevelop — APPROVED** [review]

**Review by vinit717 — APPROVED** [review]

**Pull request merged**